proposals/0000-fine-grained-undecidable.md

```diff
@@ -0,0 +1,185 @@
+
+- `UndecidableSuperClasses` omits the cycle check for all class declarations in a module. This allows for arbitrary predicates in a superclass constraint.
+- `UndecidableInstances` omits the termination check for all local quantified constraints in a module.
+- `UndecidableInstances` skips the termination check for class instances.
+- `UndecidableInstances` skips the termination check for type family equations (both in open type family instances and closed type family equations).
+
+There are also tweaks to the coverage conditions for coverage / injectivity in type families and functional dependencies.
+
+- `UndecidableInstances` allows a liberalization of the check for injective type family consistency.
+- `UndecidableInstances` allows an instance of a class with functional dependencies to use the liberal coverage condition instead of the coverage condition.
+
+## Proposed Change Specification
+
+We introduce new modifiers `%NoTerminationCheck`, `%LiberalCoverage`, `%LiberalInjectivity`, `%Overlapping`, `%Overlappable`, and `%Overlaps` which are used _before_ the keyword introducing the definition where we want to lift the corresponding restriction. The modifier syntax is taken from this [approved proposal](https://github.com/ghc-proposals/ghc-proposals/pull/370), with the additional [change](https://github.com/ghc-proposals/ghc-proposals/pull/392) of modifiers appearing _before_ the modified thing.
+
+1. Putting `%NoTerminationCheck` before a class declaration skips its superclass-cycle check.
+
+    ```haskell
+    %NoTerminationCheck ; class C (F a) => C a where ...
+    ```
+
+    ```diff
+    +cls_mod  : '%NoTerminationCheck'
+    +cls_mods : {- empty -} | cls_mods cls_mod
+
+    cl_decl :
+    - : 'class' tycl_hdr fds where_cls
+    + : cls_mods ';' 'class' tycl_hdr fds where_cls
+    ```
+
+2. Putting `%NoTerminationCheck` before the forall in a quantified constraint skips its termination check. 
+
+    ```haskell
+    f :: forall a. (%NoTerminationCheck forall b. C b a => C a a) => a -> a
+
+    -- Note the empty forall to have a target for the modifier
+    g :: forall a. (%NoTerminationCheck forall. D a a => C a) => a -> a
+    ```
+
+    ```diff
+    +forall_mod  : '%NoTerminationCheck'
+    +forall_mods : {- empty -} | forall_mods forall_mod
+
+    forall_telescope : 
+    - :             'forall' tv_bndrs '.'
+    + : forall_mods 'forall' tv_bndrs '.'
+    - |             'forall' tv_bndrs '->'
+    + | forall_mods 'forall' tv_bndrs '->'
+
+3. Putting `%NoTerminationCheck` before an instance declaration skips its termination check. Putting `%LiberalCoverage` allows for the more liberal coverage condition.
+
+    ```haskell
+    %NoTerminationCheck ; instance Eq (Tree a a) => Eq (Rose a) where ..
+    ```
+
+5. Putting `%NoTerminationCheck` within a type instance (or type in an associated type instance) skips the equation's termination check. Putting `%LiberalInjectivity` allows for the more liberal injectivity consistency check.
```

Comment:
What does "within" mean here?

Also you are talking about open type families here; mention that.

In fact, for each of these numbered items could you start with a heading.  Like

5. **Open type families**.   Putting...

That would make the structure much clearer